lib/src/main/kotlin/com/swmansion/starknet/provider/gateway/GatewayProvider.kt

```diff
@@ -53,7 +60,47 @@ class GatewayProvider(
             put("signature", JsonArray(emptyList()))
         }
 
-        return HttpRequest(Payload(url, "POST", params, body), CallContractResponse.serializer(), httpService)
+        return HttpRequest(
+            Payload(url, "POST", params, body),
+            buildHttpResponseDeserializer(CallContractResponse.serializer()),
+            httpService,
+        )
+    }
+
+    @Serializable
+    private data class GatewayError(
+        @SerialName("status_code")
+        val code: Int,
+
+        @SerialName("message")
+        val message: String,
+    )
+
+    private fun handleResponseError(response: HttpResponse): String {
+        if (!response.isSuccessful && response.body != null) {
+            try {
+                val deserializedError = Json.decodeFromString(GatewayError.serializer(), response.body)
+                throw GatewayRequestFailedException(deserializedError.code, deserializedError.message)
+            } catch (e: SerializationException) {
+                throw RequestFailedException(response.body)
+            }
+        }
+
+        if (!response.isSuccessful || response.body == null) {
+            // FIXME add more descriptive error
+            throw RequestFailedException("Request failed")
+        }
```

Comment:
Use a better error and you can move it to the top. This way this becomes a guard that ensures the response is successful and there is a body.

Reply:
I've added this FIXME myself but I now wonder if it's possible to be more descriptive in this case. The if at the beginning of this method handles cases where request failed and there's a body to work with. This one is more like a fallback.

Reply:
I think we should handle this at higher level, in HttpService itself and response would always have a non-null body. In case body == null we can't handle it specific to service anyway and it leaks into providers.